grpc/stream.go

```diff
@@ -0,0 +1,325 @@
+	builtinMetrics *metrics.BuiltinMetrics
+	obj            *goja.Object // the object that is given to js to interact with the stream
+
+	closeSend chan struct{}
+	done      chan struct{}
+
+	writeQueueCh chan message
+
+	eventListeners *eventListeners
+}
+
+// defineStream defines the goja.Object that is given to js to interact with the Stream
+func defineStream(rt *goja.Runtime, s *stream) {
+	must(rt, s.obj.DefineDataProperty(
+		"on", rt.ToValue(s.on), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+
+	must(rt, s.obj.DefineDataProperty(
+		"write", rt.ToValue(s.write), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+
+	must(rt, s.obj.DefineDataProperty(
+		"end", rt.ToValue(s.end), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+}
+
+func (s *stream) beginStream() error {
+	req := &grpcext.StreamRequest{
+		Method:           s.method,
+		MethodDescriptor: s.methodDescriptor,
+	}
+
+	stream, err := s.client.conn.NewStream(s.vu.Context(), *req, metadata.MD{})
+	if err != nil {
+		return fmt.Errorf("failed to create a new stream: %w", err)
+	}
+	s.stream = stream
+
+	go s.loop()
+
+	return nil
+}
+
+func (s *stream) loop() {
+	ctx := s.vu.Context()
+	wg := new(sync.WaitGroup)
+
+	defer func() {
+		wg.Wait()
+		s.tq.Close()
+	}()
+
+	// read & write data from/to the stream
+	wg.Add(2)
+	go s.readData(wg)
+	go s.writeData(wg)
+
+	ctxDone := ctx.Done()
+	for {
+		select {
+		case <-ctxDone:
+			// VU is shutting down during an interrupt
+			// stream events will not be forwarded to the VU
+			s.tq.Queue(func() error {
+				return s.closeWithError(nil)
+			})
+			return
+		case <-s.done:
+			return
+		}
+	}
+}
+
+func (s *stream) queueMessage(msg map[string]interface{}) {
+	s.tq.Queue(func() error {
+		rt := s.vu.Runtime()
+		listeners := s.eventListeners.all(eventData)
+
+		for _, messageListener := range listeners {
+			if _, err := messageListener(rt.ToValue(msg)); err != nil {
+				// TODO(olegbespalov) consider logging the error
+				_ = s.closeWithError(err)
+
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+// readData reads data from the stream and forward them to the readDataChan
+func (s *stream) readData(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for {
+		msg, err := s.stream.ReceiveConverted()
+
+		if err != nil && !isRegularClosing(err) {
+			s.logger.WithError(err).Debug("error while reading from the stream")
+
+			s.tq.Queue(func() error {
+				return s.closeWithError(err)
+			})
+
+			return
+		}
+
+		if len(msg) == 0 && isRegularClosing(err) {
+			s.logger.WithError(err).Debug("stream is cancelled/finished")
+
+			s.tq.Queue(func() error {
+				return s.closeWithError(nil)
+			})
+
+			return
+		}
+
+		s.queueMessage(msg)
+	}
+}
+
+func isRegularClosing(err error) bool {
+	return errors.Is(err, io.EOF) || errors.Is(err, grpcext.ErrCanceled)
+}
+
+// writeData writes data to the stream
+func (s *stream) writeData(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	writeChannel := make(chan message)
+
+	wg.Add(1)
```

Comment:
We should do the same with this.

Reply:
I believe that place is fine since it's a bit different from the https://github.com/grafana/xk6-grpc/pull/1/files#r1160597793 where we could end up with a case where `writeData` will be never executed. 

@mstoykov please correct me if I wrong

Reply:
Yes - here `wg.Add` was already called for `writeData` and deferred `wg.Done` has not yet ran, so any `wg.Wait` will wait for this as well.